Library/Formula/z3.rb

```diff
@@ -12,31 +19,110 @@ class Z3 < Formula
+      args = ["--prefix=#{prefix}"]
+      args << "-j" if build.with? "java"
+      args << "--ml" if build.with? "ocaml"
+      system python, "scripts/mk_make.py", *args
       cd "build" do
         system "make"
         system "make", "install"
+
+        if build.with? "ocaml"
+          (lib/"ocaml").install Dir["api/ml/*"]
+        end
       end
     end
 
     pkgshare.install "examples"
   end
 
+  def caveats
+    s = ""
+    if build.with? "java"
+      s += <<-EOS.undent
+
+        Java bindings were installed into #{libexec}/com.microsoft.z3.jar and #{lib}/z3java.dylib.
+    To use these, set the classpath and java.library.path accordingly.
+        Alternatively, you may need to link the Java bindings into the Java Extensions folder:
+          sudo mkdir -p /Library/Java/Extensions
+          sudo ln -s #{lib}/libz3java.dylib #{libexec}/com.microsoft.z3.jar /Library/Java/Extensions
+      EOS
+    end
+    s
+  end
+
   test do
+    # Test C bindings
     system ENV.cc, "-I#{include}", "-L#{lib}", "-lz3",
            pkgshare/"examples/c/test_capi.c", "-o", testpath/"test"
-    system "./test"
+    assert_match(/Z3 #{version}/, shell_output("./test"))
+
+    # Test z3 binary with SMT2 input
+    field_test = testpath/"test_field.smt"
+    field_test.write <<-EOS.undent
+      (declare-const a Real)
+      (declare-const b Real)
+      (declare-const c Real)
+
+      ; Verify the axioms of the field of Real numbers
+      (assert (= (+ (+ a b) c) (+ a (+ b c))))        ; Associativity of addition
+      (assert (= (+ a b) (+ b a)))                    ; Commutativity of addition
+      (assert (= (+ (- 0 a) a) 0))                    ; Existence of an inverse element of addition
+      (assert (= (+ 0 a) a))                          ; Existence of a neutral element of addition
+      (assert (= (* (* a b) c) (* a (* b c))))        ; Associativity of multiplication
+      (assert (= (* a b) (* b a)))                    ; Commutativity of multiplication
+      (assert (= (* (/ 1 a) a) 0))                    ; Existence of an inverse element of addition
+      (assert (= (* 1 a) a))                          ; Existence of a neutral element of addition
+      (assert (= (* a (+ b c)) (+ (* a b) (* a c))))  ; Left-distributive property
+      (assert (= (* (+ a b) c) (+ (* a c) (* b c))))  ; Right-distributive property
+
+      (check-sat)
+    EOS
+    field_sat = shell_output("#{bin}/z3 -smt2 #{field_test}")
+    assert_equal "sat\n", field_sat
+
+    # Test if a simple unsatisfiable test fails
+    unsat_test = testpath/"unsat_test.smt"
+    unsat_test.write <<-EOS.undent
+      (assert (= 42 23))
+      (check-sat)
+    EOS
+    unsat = shell_output("#{bin}/z3 -smt2 #{unsat_test}")
+    assert_equal "unsat\n", unsat
+
+    if build.with? "python"
+      assert_equal "#{version}\n", shell_output('python -c "import z3; print z3.get_version_string()"')
+    end
+
+    if build.with? "java"
```

Comment:
Probably, for now. We're considering our options on er, options, but it can be quite confusing and difficult for us if people report failing tests on non-default options and then never give us more information to debug, which happens frequently sadly.